File: 조준장/level2/괄호 변환.py
def 올바른_문자열인지_확인(p):
    stack = []
    for i in  p:
        if stack and stack[-1] == "(" and i == ")" :
            stack.pop()
        else:
            stack.append(i)
    
    if stack:
        return False
    return True

def u와v_분리(p):
    left = 0
    right = 0
    for i in range(len(p)):
        if p[i] == ")":
            left += 1
        else:
            right += 1
        
        if left == right:
            # u, v 분리
            u = p[:i+1]
            v = p[i+1:]
            
            if 올바른_문자열인지_확인(u): return u + u와v_분리(v)
            return "(" + u와v_분리(v) + ")" + 올바른_문자열_만들기(u) 
        
    return ""

def 올바른_문자열_만들기(u):
    print(u.sprit())
    print(u.lstrip().rstrip())
    return 괄호_방향_뒤집기(u.lstrip().rstrip())

def 괄호_방향_뒤집기(u):
    U = ""
    for i in u:
        if i == ")":
            U += "("
        else:
            U += ")"
    return U

def solution(p):
    if 올바른_문자열인지_확인(p): return p
    return u와v_분리(p)

File: 조준장/level2/괄호변환.kt
import java.util.*

class Solution {
    fun solution(p: String): String {
        var answer = ""
        
        return if (`올바른 괄호인지 확인`(p)) p
        else `u와 v로 분리`(p)
    }

    fun `u와 v로 분리`(w : String) : String {
        var left = 0
        var right = 0
        repeat(w.length) {
            if (w[it] == '(') {
                left++
            } else {
                right++
            }
            
            if(left == right) {
                val u = w.substring(0, it + 1)
                val v = w.substring(it + 1, w.length)
                
                if (`올바른 괄호인지 확인`(u)) return u + `u와 v로 분리`(v)
                return "(" + `u와 v로 분리`(v) + ")" +`올바른 문자열 만들기`(u)
            }
        }
        return ""
    }
    
    fun `올바른 괄호인지 확인`(v : String) : Boolean {
        val stack = Stack<String>()
        v.forEach {
            if (stack.isNotEmpty() && stack.last() == "(" && it == ')') {
                stack.pop()
            } else {
                stack.push(it.toString())
            }
        }
        
        return if (stack.isEmpty()) true else false
    }
    
    fun `올바른 문자열 만들기`(u : String) : String {
        var newU = u
        if (newU.length >= 1) {
            newU = newU.drop(1)
        }
        if (newU.length >= 1) {
            newU = newU.dropLast(1)
        }

        var str = ""
        
        newU.forEach {
            if (it == '(') {
                str += ")"
            } else {
                str += "("
            }
        }
        
        return str
    }
}
